src/realm/views.cpp

```diff
@@ -190,9 +199,121 @@ bool SortDescriptor::Sorter::operator()(IndexPair i, IndexPair j, bool total_ord
     return total_ordering ? i.index_in_view < j.index_in_view : 0;
 }
 
-void RowIndexes::do_sort(const SortDescriptor& order, const SortDescriptor& distinct)
+void DescriptorOrdering::append_sort(SortDescriptor sort)
+{
+    size_t orig_length = m_descriptors.size();
+    if (orig_length == 0) {
+        m_descriptors.emplace_back(std::move(sort));
+    }
+    else if (descriptor_is_sort(orig_length - 1)) {
+        // merge them together by appending internally
+        m_descriptors[orig_length - 1].merge_with(std::move(sort));
+    }
+    else { // last descriptor is distinct, just add it
+        m_descriptors.emplace_back(std::move(sort));
+    }
+}
+
+void DescriptorOrdering::append_distinct(DistinctDescriptor distinct)
+{
+    size_t orig_length = m_descriptors.size();
+    if (orig_length == 0) {
+        // add an empty sort descriptor first to keep ordering
+        m_descriptors.emplace_back(SortDescriptor());
+        m_descriptors.emplace_back(std::move(distinct));
+    }
+    else if (descriptor_is_distinct(orig_length - 1)) {
+        // add the next distinct operation, separated by a placeholder sort
+        m_descriptors.emplace_back(SortDescriptor());
+        m_descriptors.emplace_back(std::move(distinct));
+    }
+    else { // last descriptor is sort, just add it
+        m_descriptors.emplace_back(std::move(distinct));
+    }
+}
+
+bool DescriptorOrdering::descriptor_is_sort(size_t index) const
+{
+    // The list is kept in alternating order of sort/distinct/sort/distinct...
```

Comment:
This feels like a strange choice. Would it not be simpler to have each `SortDescriptor` know whether it's for sorting or distinct, and do away with using even and odd indexes and the placeholder descriptors?

Ideally `SortDescriptor` and `DistinctDescriptor` would be different types, with `DescriptorOrdering` holding something like a `std::vector<std::variant<SortDescriptor, DistinctDescriptor>>` rather than trying to have `SortDescriptor` serve both purposes, but we don't have an implementation of `std::variant` available to us right now so we'll have to punt on that.